cvmfs/sqlitemem.cc

```diff
@@ -0,0 +1,508 @@
+}
+
+
+/**
+ * Sqlite ensures that ptr != NULL.
+ */
+void MemoryManager::xFree(void *ptr) {
+  instance_->PutMemory(ptr);
+}
+
+
+/**
+ * Sqlite ensures that ptr != NULL and new_size > 0.
+ */
+void *MemoryManager::xRealloc(void *ptr, int new_size) {
+  void *new_ptr = xMalloc(new_size);
+  memcpy(new_ptr, ptr, new_size);
+  xFree(ptr);
+  return new_ptr;
+}
+
+
+/**
+ * Sqlite ensures that ptr != NULL.
+ */
+int MemoryManager::xSize(void *ptr) {
+  return instance_->GetMemorySize(ptr);
+}
+
+
+int MemoryManager::xRoundup(int size) {
+  return size;
+}
+
+
+int MemoryManager::xInit(void *app_data __attribute__((unused))) {
+  return SQLITE_OK;
+}
+
+
+void MemoryManager::xShutdown(void *app_data __attribute__((unused))) {
+}
+
+
+void MemoryManager::AssignGlobalArenas() {
+  if (assigned_) return;
+  int retval;
+
+  retval = sqlite3_config(SQLITE_CONFIG_SCRATCH, scratch_memory_,
+                          kScratchSlotSize, kScratchNoSlots);
+  assert(retval == SQLITE_OK);
+
+  retval = sqlite3_config(SQLITE_CONFIG_PAGECACHE, page_cache_memory_,
+                          kPageCacheSlotSize, kPageCacheNoSlots);
+  assert(retval == SQLITE_OK);
+
+  retval = sqlite3_config(SQLITE_CONFIG_GETMALLOC, &sqlite3_mem_vanilla_);
+  assert(retval == SQLITE_OK);
+  retval = sqlite3_config(SQLITE_CONFIG_MALLOC, &mem_methods_);
+  assert(retval == SQLITE_OK);
+
+  assigned_ = true;
+}
+
+
+/**
+ * Needs to be the first operation on an opened sqlite database.  Returns the
+ * location of the buffer.
+ */
+void *MemoryManager::AssignLookasideBuffer(void *db) {
+  MutexLockGuard lock_guard(lock_);
+
+  sqlite3 *sqlite_db = reinterpret_cast<sqlite3 *>(db);
+  void *buffer = GetLookasideBuffer();
+  assert(buffer != NULL);
+  int retval = sqlite3_db_config(sqlite_db, SQLITE_DBCONFIG_LOOKASIDE,
+    buffer, kLookasideSlotSize, kLookasideSlotsPerDb);
+  assert(retval == SQLITE_OK);
+  return buffer;
+}
+
+
+void MemoryManager::CleanupInstance() {
+  delete instance_;
+  instance_ = NULL;
+}
+
+
+/**
+ * Opens a new arena if necessary.
+ */
+void *MemoryManager::GetLookasideBuffer() {
+  void *result;
+  vector<LookasideBufferArena *>::reverse_iterator reverse_iter =
+    lookaside_buffer_arenas_.rbegin();
+  vector<LookasideBufferArena *>::reverse_iterator i_rend =
+    lookaside_buffer_arenas_.rend();
+  for (; reverse_iter != i_rend; ++reverse_iter) {
+    result = (*reverse_iter)->GetBuffer();
+    if (result != NULL)
+      return result;
+  }
+
+  LookasideBufferArena *new_arena = new LookasideBufferArena();
+  lookaside_buffer_arenas_.push_back(new_arena);
+  return new_arena->GetBuffer();
+}
+
+
+int MemoryManager::GetMemorySize(void *ptr) {
+  return MallocArena::GetMallocArena(ptr)->GetSize(ptr);
+}
+
+
+/**
+ * Opens new arenas as necessary.
+ */
+void *MemoryManager::GetMemory(int size) {
+  void *p = malloc_arenas_[idx_last_arena_]->Malloc(size);
+  if (p != NULL)
+    return p;
+  unsigned N = malloc_arenas_.size();
```

Comment:
Knit picking: newlines to separate code blocks.

1. Try last successful arena
[newline]
2. Try all other arenas
[newline]
3. Create new arena